lib/mongoid/query_cache.rb

```diff
@@ -224,34 +223,39 @@ module View
       #
       # @since 5.0.0
       def each
-        if system_collection? || !QueryCache.enabled?
+        if system_collection? || !QueryCache.enabled? || (respond_to?(:write?, true) && write?)
           super
         else
-          unless cursor = cached_cursor
-            read_with_retry do
-              server = server_selector.select_server(cluster)
-              result = send_initial_query(server)
-              if result.cursor_id == 0 || result.cursor_id.nil?
-                cursor = CachedCursor.new(view, result, server)
-                QueryCache.cache_table[cache_key] = cursor
-              else
-                cursor = Mongo::Cursor.new(view, result, server)
+          @cursor = nil
+          unless @cursor = cached_cursor
+
+            if driver_supports_cursor_sessions?
+              session = client.send(:get_session, @options)
+              read_with_retry(session, server_selector) do |server|
+                result = send_initial_query(server, session)
+                @cursor = cursor(result, server, session)
+              end
+            else
+              read_with_retry do
+                server = server_selector.select_server(cluster)
+                result = send_initial_query(server)
+                @cursor = cursor(result, server)
```

Comment:
I do not love this code 😆 If you have suggestions about how to improve it, I am open to hearing them.

Reply:
Can the method be called something other than `cursor` to avoid future confusion? Maybe `get_cursor` even?

Reply:
Sure thing.